### PR TITLE
Jmx Reporter Addition to allow for New Relic over Jolokia alternative

### PR DESCRIPTION
**Pull request by evh69**
This pull request is response to https://stackoverflow.com/questions/51087910/jmx-metrics-and-new-relic.

The [Corda #monitoring your node](https://docs.corda.net/head/node-administration.html#monitoring-your-node) suggests, "Cloud metrics services like New Relic also understand JMX, typically, by providing their own agent that uploads the data to their service on a regular schedule."

As an alternative to the OOTB Jolokia configuration, this pull request introduces an initial solution and simple mechanism for allowing the node administrator the option of including a New Relic Reporter over the Jolokia reporter.

By setting jmxReporterType to NEW_RELIC in the node.conf file, the Node.kt startup() will derive that the [NewRelic](https://metrics.dropwizard.io/3.2.3/manual/third-party.html) _NewRelicReporter_ third [party dropwizard library](https://github.com/palominolabs/metrics-new-relic) be used instead of the default configuration that uses the Jolokia JxmReporter for reporting metrics.

The assumption is that if jmxReporterType is set to NEW_RELIC the node administrator:

1. Has a registered account with New Relic
2. Follows the [New Relic Java Agent Configuration](https://docs.newrelic.com/docs/agents/java-agent/configuration/java-agent-configuration-config-file) to install and configure the java agent newrelic.jar.

In testing, the new relic jar can be placed in nodes/[NODE_NAME}/driver folder.

Then simply replace -javaagent:drivers/jolokia-jvm-1.3.7-agent.jar with -javaagent:drivers/newrelic.jar when one executes java -jar corda.jar.

### CODE REVIEW

**Comment by tlil** [comment]
Hi @evh69 - thanks for your contribution. I will review this later. In the meantime, could you please resolve the conflicts in `Node.kt`?

**Comment by mikehearn** [comment]
Thanks for the contribution. Can you state that you are in compliance with the developer certificate of origin (see CONTRIBUTING.md)? A simple affirmative statement is sufficient.

**Comment by evh69** [comment]
@mikehearn affirmative, nothing broken, at least prior to the resolving conflicts - 👍 ...

@tlil - I will have a look later today.

**Comment by evh69** [comment]
@tlil - Done.

**Comment by tlil** [comment]
@evh69 - to your message on Slack last night, I've kicked off the unit and integration tests again as they failed due to flakiness (nothing to do with this PR). Will review and get back to you on the PR itself today..

**Comment by evh69** [comment]
@tlil ... Great ... As an initial pass, i think it will work.  In spring speak ... Realizing, there is not any spring in the mix ... I would like to actually do something more spring like, making the Reporter an @Service and allowing the application team to wire the right bean on the Corda main config.  But, I don't see any of that type of thing going on.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 1]
```diff
@@ -406,6 +402,47 @@ open class Node(configuration: NodeConfiguration,
...
+
+    private fun registerNewRelicReporter (registry: MetricRegistry) {
+        staticLog.info("Registering New Relic JMX reporter:")
+        val reporter = NewRelicReporter.forRegistry(registry)
+                .name("new relic reporter")
+                .filter(MetricFilter.ALL)
+                .attributeFilter(AllEnabledMetricAttributeFilter())
+                .rateUnit(TimeUnit.SECONDS)
+                .durationUnit(TimeUnit.MILLISECONDS)
+                .metricNamePrefix("CorDapp/")
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: Would a different metric name prefix make more sense? E.g. just `"corda/"` ? I'm just thinking for more generic metrics that might get introduced and which pertains to the platform rather than a specific CorDapp.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 2]
```diff
@@ -406,6 +402,47 @@ open class Node(configuration: NodeConfiguration,
...
+                ObjectName("$domain:name=$category")
+            else
+                ObjectName("$domain:type=$category,name=$subName")
+        }.build().start()
+    }
+
+    private fun registerNewRelicReporter (registry: MetricRegistry) {
+        staticLog.info("Registering New Relic JMX reporter:")
+        val reporter = NewRelicReporter.forRegistry(registry)
+                .name("new relic reporter")
```
Comment (tlil) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: Capitalise? `"New Relic Reporter"`

File: node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt [hunk 3]
```diff
@@ -208,6 +208,21 @@ class NodeConfigurationImplTest {
         assertEquals(compatibilityZoneURL, configuration.networkServices!!.networkMapURL)
     }
 
+    @Test
```
Comment (tlil) [inline node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt]: Could you add an explicit test for `jmxReporterType is null and defaults to Jokolia` as well?

File: node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt [hunk 4]
```diff
@@ -208,6 +208,21 @@ class NodeConfigurationImplTest {
...
+    fun `jmxReporterType is not null and defaults to Jokolia`() {
+
+        var rawConfig = getConfig("working-config.conf", ConfigFactory.parseMap(mapOf("devMode" to true)))
+        val nodeConfig = rawConfig.parseAsNodeConfiguration()
+        assertTrue(JmxReporterType.JOLOKIA.toString() == nodeConfig.jmxReporterType.toString())
+    }
+
+    @Test
+    fun `jmxReporterType is not null and defaults to New Relic`() {
+        var rawConfig = ConfigFactory.parseResources("test-working-config-newrelic.conf", ConfigParseOptions.defaults().setAllowMissing(false))
```
Comment (tlil) [inline node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt]: You could also use something like:

```kt
rawConfig = rawConfig.withValue("jmxReporterType", ConfigValueFactory.fromAnyRef("NEW_RELIC"))
```

That way you'd avoid introducing a new config file. :-)

File: docs/source/corda-configuration-file.rst [hunk 5]
```diff
@@ -245,6 +245,12 @@ absolute path to the node's base directory.
 
 :flowMonitorSuspensionLoggingThresholdMillis: Threshold ``Duration`` suspended flows waiting for IO need to exceed before they are logged. Default value is ``60 seconds``.
 
+:jmxReporterType: Optionally set, if excluded the NodeConfiguration will default to a value JOLOKIA.  Provides an option for registering an alternative Jmx Reporter.  Acceptable values are
```
[the remark this answers is not in the document]
Comment (tlil) [reply docs/source/corda-configuration-file.rst]: Perhaps rephrase to something like:

>Provides an option for registering an alternative JMX reporter.  Available options are ``JOLOKIA`` and ``NEW_RELIC``. If no value is provided, ``JOLOKIA`` will be used.

File: docs/source/corda-configuration-file.rst [hunk 6]
```diff
@@ -245,6 +245,12 @@ absolute path to the node's base directory.
...
+    JOLOKIA and NEW_RELIC
+
+                    .. note:: The Jolokia configuraiton is provided by default.  The NewRelic configuration leverages https://metrics.dropwizard.io/3.2.3/manual/third-party.html NewRelicReporter solution.
```
Comment (tlil) [inline docs/source/corda-configuration-file.rst]: Spelling mistake: `configuration`

Comment (tlil) [reply docs/source/corda-configuration-file.rst]: NewRelic => New Relic

Comment (tlil) [reply docs/source/corda-configuration-file.rst]: Can you turn these two links into named hyperlinks? See the following documentation for more information: http://docutils.sourceforge.net/docs/user/rst/quickref.html#external-hyperlink-targets

File: docs/source/corda-configuration-file.rst [hunk 7]
```diff
@@ -245,6 +245,12 @@ absolute path to the node's base directory.
...
+                            See https://docs.newrelic.com/docs/agents/java-agent/getting-started/introduction-new-relic-java for details on getting started, and installing the New Relic java agent.
```
Comment (tlil) [inline docs/source/corda-configuration-file.rst]: java agent => Java Agent

**Comment by tlil** [comment]
Regarding:
>@tlil ... Great ... As an initial pass, i think it will work. In spring speak ... Realizing, there is not any spring in the mix ... I would like to actually do something more spring like, making the Reporter an @service and allowing the application team to wire the right bean on the Corda main config. But, I don't see any of that type of thing going on.

Please just keep this as it is. We don't use Spring in the core part of the platform at the moment.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt (hunk 1 above)
Comment (evh69) [reply node/src/main/kotlin/net/corda/node/internal/Node.kt]: A default of corda is fine for now.  I actually debated including a configuration value for all of the options.  But, decided to wait, this could come later.

File: docs/source/corda-configuration-file.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (evh69) [reply docs/source/corda-configuration-file.rst]: Done.

File: node/build.gradle [hunk 8]
```diff
@@ -189,6 +189,8 @@ dependencies {
 
     // Jolokia JVM monitoring agent, required to push logs through slf4j
     compile "org.jolokia:jolokia-jvm:${jolokia_version}:agent"
+    // https://mvnrepository.com/artifact/com.palominolabs.metrics/metrics-new-relic
```
[the remark this answers is not in the document]
Comment (evh69) [reply node/build.gradle]: Done.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 9]
```diff
@@ -406,6 +402,47 @@ open class Node(configuration: NodeConfiguration,
...
+    fun registerJmxReporter(metrics: MetricRegistry) {
+        staticLog.info("Registering JMX reporter:")
+        when (configuration.jmxReporterType) {
+            JmxReporterType.JOLOKIA -> registerJolokiaReporter(metrics)
+            JmxReporterType.NEW_RELIC -> registerNewRelicReporter(metrics)
+        }
+    }
+
+    private fun registerJolokiaReporter(registry: MetricRegistry) {
+        staticLog.info("Registering Jolokia JMX reporter:")
```
[the remark this answers is not in the document]
Comment (evh69) [reply node/src/main/kotlin/net/corda/node/internal/Node.kt]: Done.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt (hunk 1 above)
Comment (evh69) [reply node/src/main/kotlin/net/corda/node/internal/Node.kt]: Done.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt (hunk 2 above)
Comment (evh69) [reply node/src/main/kotlin/net/corda/node/internal/Node.kt]: Done.

File: node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt (hunk 3 above)
Comment (evh69) [reply node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt]: Done.

File: node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt (hunk 4 above)
Comment (evh69) [reply node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationImplTest.kt]: Done.  Thx. for the tip.  Deleted extra config files.

File: docs/source/corda-configuration-file.rst (hunk 6 above)
Comment (evh69) [reply docs/source/corda-configuration-file.rst]: Done.

File: docs/source/corda-configuration-file.rst (hunk 7 above)
Comment (evh69) [reply docs/source/corda-configuration-file.rst]: Done.

File: docs/source/corda-configuration-file.rst [hunk 10]
```diff
@@ -245,11 +245,12 @@ absolute path to the node's base directory.
...
 
-:jmxReporterType: Optionally set, if excluded the NodeConfiguration will default to a value JOLOKIA.  Provides an option for registering an alternative Jmx Reporter.  Acceptable values are
-    JOLOKIA and NEW_RELIC
+:jmxReporterType:  Provides an option for registering an alternative JMX reporter. Available options are ``JOLOKIA`` and ``NEW_RELIC``. If no value is provided, ``JOLOKIA`` will be used.
 
-                    .. note:: The Jolokia configuraiton is provided by default.  The NewRelic configuration leverages https://metrics.dropwizard.io/3.2.3/manual/third-party.html NewRelicReporter solution.
-                            See https://docs.newrelic.com/docs/agents/java-agent/getting-started/introduction-new-relic-java for details on getting started, and installing the New Relic java agent.
+                    .. note:: The Jolokia configuration is provided by default.  The New Relic configuration leverages the Dropwizard_ NewRelicReporter solution. See Introduction-new-relic-java_  for details on getting started, and installing the New Relic Java Agent.
```
Comment (tlil) [inline docs/source/corda-configuration-file.rst]: The formatting of this now looks fine (see screenshot below):

image

However, the links are invalid. E.g., it generates:

```html
<a class="reference external" href="&lt;https://docs.newrelic.com/docs/agents/java-agent/getting-started/introduction-new-relic-java/&gt;">Introduction-new-relic-java</a>
```

Can you fix that please, @evh69 ? And perhaps rephrase to something like:

>See [Introduction to New Relic for Java]() for details on how to get started and how to install the New Relic Java agent.

Comment (evh69) [reply docs/source/corda-configuration-file.rst]: Arg!!! Sorry, I failed to try the links.  I am using http://rst.ninjs.org/ to view the markdown results.   As, intellij does not appear to have a good plugin.  And, the results looked good.  I have adjusted and committed.

Comment (tlil) [reply docs/source/corda-configuration-file.rst]: @evh69 No worries. Thanks. 

Can you make the following two updates as well please?

>See Introduction to New Relic for Java for details on **how to get** started and how to install the New Relic Java agent.

i.e., change from "on getting started" to "on how to get started"

And, change "Java Agent" to "Java agent".

Thanks
